lib/audit/catalog_to_moab.rb

```diff
@@ -5,18 +5,14 @@
 class CatalogToMoab
 
   # allows for sharding/parallelization by storage_dir
-  def self.check_version_on_dir_of_batch(last_checked_b4_date, storage_dir, limit)
+  def self.check_version_on_dir(last_checked_b4_date, storage_dir, limit)
     # TODO: ensure last_checked_version_b4_date is in the right format for query - see #485
-    pcs = PreservedCopy.least_recent_version_audit(last_checked_b4_date, storage_dir).limit(limit)
-    pcs.each do |pc|
-      c2m = CatalogToMoab.new(pc, storage_dir)
-      c2m.check_catalog_version
-    end
-  end
-
-  def self.check_version_on_dir(last_checked_b4_date, storage_dir)
-    unless PreservedCopy.least_recent_version_audit(last_checked_b4_date, storage_dir).count.zero?
-      check_version_on_dir_of_batch(last_checked_b4_date, storage_dir, Settings.c2m_sql_limit)
+    pcs = PreservedCopy.least_recent_version_audit(last_checked_b4_date, storage_dir)
```

Comment:
once again, please pardon my ignorance here.

Based on what @jmartin-sul said in the earlier PR https://github.com/sul-dlss/preservation_catalog/pull/519#discussion_r165276153 and F2F at standup:

- doing a "count" query is cheap
- getting back a LOT of actual objects is expensive

So is this a step in the wrong direction?

Not sure if the following s expensive or not:
- "order by" on a large set of objects (order by date)


The following could avoid repeating the count query (which probably does NOT matter):

```ruby
num_to_process = PreservedCopy.least_recent_version_audit(last_checked_b4_date, storage_dir).count
while num_to_process > 0
  check_version_on_dir_of_batch(last_checked_b4_date, storage_dir, Settings.c2m_sql_limit)
  num_to_process = num_to_process - Settings.c2m_sql_limit
end
```

If @jmartin-sul agrees, I think leaving the code as it was will likely have better performance.  I would guess the performance hits are in sorting a huge number of db entries, in retrieving a huge number of db entries, and in instantiating ActiveRecord objects.

Reply:
ah, ok, some clarification:
* my understanding, not empirically tested here, but which i have some confidence in, is that count queries are indeed cheaper.
  * but, my guess without yet having tried it in a console is that 
   ```ruby
    num_to_process = PreservedCopy.least_recent_version_audit(last_checked_b4_date, storage_dir).count
    while num_to_process > 0
    ```
   will loop infinitely or not at all, because num_to_process will be the integer result of the `count` (i think `count` and `pluck` both cause you to run the query immediately and get a value or values back immediately, as opposed to getting an ActiveRecord Relation that executes lazily once you do something to access the result).  so you'd probably want to do the same but skip the intermediate variable.
* we shouldn't do quite what's happening on this PR now, which i believe will inadvertently load all our pres copies into memory:  i think `unless pcs.empty?` will treat the `least_recent_version_audit` `Relation` as an array, just to see whether it has any elements.  since there's no limit clause on that one, all records would be loaded, i think.  and then the query will re-run with the limit (which would likely be cheaper than the check, ironically, due to the limit).  so... there seem to me to be two options
  * what's here, minus the initial check to see whether there are elements.  just the `pcs.limit(limit).each do |pc| ...` loop.  if there are pres copies that satisfy the query, `limit` or fewer of them will be returned and iterated over, and if there are none that satisfy the query, no loop iterations occur.  no need to get a count beforehand, i don't think.
  * as naomi suggests, use the count to have a `while` loop iterate over all qualifying  pres copies in `limit` sized batches (but inline the `count` query into the loop condition, or assign it to a lambda or whatever so that it's re-evaluated every time, since `count` returns an int immediately).

i'm partial to the second, because then we don't have to get fancy with our cron job scheduling, other than maybe a lockfile.  we iterate through everything available till nothing's left to be audited.


http://api.rubyonrails.org/classes/ActiveRecord/Calculations.html#method-i-count
http://api.rubyonrails.org/classes/ActiveRecord/Calculations.html#method-i-pluck

Reply:
d'oh, sorry @ndushay, i didn't see the decrementing in your loop.  probably a safer bet than my proposal of re-running the query, since it would result in a definite finite number of iterations (whereas getting the count anew could go forever if our throughput was too low or we started failing to update audit dates).